core/trino-main/src/main/java/io/trino/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -1303,4 +1321,77 @@ public PlanNode getRoot()
             return mappings;
         }
     }
+
+    private static class DynamicFilterVisitor
+            extends PlanVisitor<PlanNode, UnaliasDynamicFilterContext>
+    {
+        private final Metadata metadata;
+
+        private DynamicFilterVisitor(Metadata metadata)
+        {
+            this.metadata = metadata;
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, UnaliasDynamicFilterContext context)
+        {
+            List<PlanNode> newSources = node.getSources().stream()
+                    .map(source -> source.accept(this, context))
+                    .collect(toImmutableList());
+
+            return replaceChildren(node, newSources);
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, UnaliasDynamicFilterContext context)
+        {
+            if (context.getDynamicFilterIdMap().isEmpty()) {
+                return node;
+            }
+            return new FilterNode(node.getId(), node.getSource(), updateDynamicFilterIds(context.getDynamicFilterIdMap(), node.getPredicate()));
+        }
+
+        private Expression updateDynamicFilterIds(Map<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap, Expression predicate)
+        {
+            List<Expression> conjuncts = extractConjuncts(predicate);
+            List<Expression> mapped = conjuncts.stream().map(conjunct -> getDescriptor(conjunct)
+                    .map(descriptor -> dynamicFilterIdMap.get(descriptor.getId()))
+                    .map(mappedId -> replaceDynamicFilterId((FunctionCall) conjunct, mappedId))
+                    .orElse(conjunct)).collect(toImmutableList());
+            return combineConjuncts(metadata, mapped);
+        }
+
+        private Expression replaceDynamicFilterId(FunctionCall dynamicFilterFunctionCall, DynamicFilterId newId)
```

Comment:
move that method to `DynamicFilters` so that whoever modifies `DynamicFilters` will know about this use case.